### PR TITLE
feat: add the ability to install plugins remotely via URI

### LINKED ISSUE #2292: 完善插件安装功能
### 你当前使用的版本

next

### 描述一下此特性

/kind feature
/milestone 2.0
/area core
/area admin

支持插件安装，支持远程链接下载和本地上传。目前虽然已经实现上传安装，但是没有 review 资源文件（`plugin.yaml` `roleTemplates.yaml`）的过程。需要实现上传之后读取资源，在前端展示资源文件以及手动创建资源。

/cc @halo-dev/sig-halo 
/cc @halo-dev/sig-halo-admin 

### 附加信息

_No response_

### PR DESCRIPTION
**Pull request by guqing**
#### What type of PR is this?
/kind feature
/area core
/area console
/milestone 2.6.x
/kind api-change

#### What this PR does / why we need it:
支持通过 URI 远程安装和升级插件

how to test it?
1. 测试插件安装
```shell
curl -u admin:admin -X POST http://localhost:8090/apis/api.console.halo.run/v1alpha1/plugins/-/install-from-uri --data '{
  "uri": "https://halo.run/apis/api.store.halo.run/v1alpha1/applications/app-KhIVw/releases/app-release-canxF/download/app-release-canxF-znFre"
}'
```
2. 测试插件升级
```shell
curl -u admin:admin -X POST http://localhost:8090/apis/api.console.halo.run/v1alpha1/plugins/PluginFeed/upgrade-from-uri --data '{
  "uri": "https://halo.run/apis/api.store.halo.run/v1alpha1/applications/app-KhIVw/releases/app-release-canxF/download/app-release-canxF-znFre"
}'
```
#### Which issue(s) this PR fixes:
Fixes #2292

#### Does this PR introduce a user-facing change?
```release-note
支持通过 URI 远程安装和升级插件
```

### CODE REVIEW

**Review by ruibaby — APPROVED** [review]
/approve

File: application/src/main/java/run/halo/app/core/extension/endpoint/PluginEndpoint.java [hunk 1]
```diff
@@ -177,6 +220,55 @@ public RouterFunction<ServerResponse> endpoint() {
             .build();
     }
 
+    private Mono<ServerResponse> upgradeFromUri(ServerRequest request) {
+        final var name = request.pathVariable("name");
+        return request.bodyToMono(UpgradeFromUriRequest.class)
+            .flatMap(upgradeRequest -> Mono.fromCallable(() -> DataBufferUtils.toInputStream(
+                reactiveUrlDataBufferFetcher.fetch(upgradeRequest.uri())))
+            )
+            .doOnError(throwable -> {
```
Comment (JohnNiang) [inline application/src/main/java/run/halo/app/core/extension/endpoint/PluginEndpoint.java:229]: 这里建议用 `reactor.core.publisher.Mono#onErrorMap(java.util.function.Function<? super java.lang.Throwable,? extends java.lang.Throwable>)`。

**Review by JohnNiang — CHANGES_REQUESTED** [review]

File: application/src/main/java/run/halo/app/core/extension/endpoint/PluginEndpoint.java [hunk 2]
```diff
@@ -508,4 +600,11 @@ private Mono<Path> transferToTemp(FilePart filePart) {
             );
     }
 
+    private Mono<Path> transferToTemp(InputStream inputStream) {
+        return Mono.fromCallable(() -> {
+            Path tempFile = Files.createTempFile("halo-plugins", ".jar");
+            FileCopyUtils.copy(inputStream, Files.newOutputStream(tempFile));
+            return tempFile;
+        });
```
Comment (JohnNiang) [inline application/src/main/java/run/halo/app/core/extension/endpoint/PluginEndpoint.java:608]: 这里涉及到阻塞操作（文件复制），建议使用 subscribeOn。

需要注意的是：主题安装那边可能也有类似的问题。

**Review by JohnNiang — CHANGES_REQUESTED** [review]

File: application/src/main/java/run/halo/app/core/extension/endpoint/PluginEndpoint.java (hunk 2 above)
Comment (guqing) [reply application/src/main/java/run/halo/app/core/extension/endpoint/PluginEndpoint.java]: good catch, For more details, refer to https://projectreactor.io/docs/core/release/reference/#faq.wrap-blocking
image

**Review by JohnNiang — APPROVED** [review]
/approve

**Review by ruibaby — APPROVED** [review]
/lgtm

**Pull request merged**